server/node/copy_from.go

```diff
@@ -0,0 +1,203 @@
+type CopyFrom struct {
+	DatabaseName string
+	TableName    doltdb.TableName
+	File         string
+	Delimiter    string
+	Null         string
+}
+
+var _ vitess.Injectable = (*CopyFrom)(nil)
+var _ sql.ExecSourceRel = (*CopyFrom)(nil)
+
+// NewCopyFrom returns a new *CopyFrom.
+func NewCopyFrom(databaseName string, tableName doltdb.TableName, fileName string) *CopyFrom {
+	return &CopyFrom{
+		DatabaseName: databaseName,
+		TableName:    tableName,
+		File:         fileName,
+		Delimiter:    "\t",
+		Null:         `\N`,
+	}
+}
+
+// CheckPrivileges implements the interface sql.ExecSourceRel.
+func (cf *CopyFrom) CheckPrivileges(ctx *sql.Context, opChecker sql.PrivilegedOperationChecker) bool {
+	// https://www.postgresql.org/docs/15/sql-copy.html
+	// ... database superusers or users who are granted one of the roles `pg_read_server_files`, `pg_write_server_files`, or `pg_execute_server_program` ...
+	return true
+}
+
+// Children implements the interface sql.ExecSourceRel.
+func (cf *CopyFrom) Children() []sql.Node {
+	return nil
+}
+
+// IsReadOnly implements the interface sql.ExecSourceRel.
+func (cf *CopyFrom) IsReadOnly() bool {
+	return false
+}
+
+// Resolved implements the interface sql.ExecSourceRel.
+func (cf *CopyFrom) Resolved() bool {
+	return true
+}
+
+// RowIter implements the interface sql.ExecSourceRel.
+func (cf *CopyFrom) RowIter(ctx *sql.Context, r sql.Row) (_ sql.RowIter, err error) {
+	// Fetch the table
+	table, err := core.GetSqlTableFromContext(ctx, cf.DatabaseName, cf.TableName)
+	if err != nil {
+		return nil, err
+	}
+	if table == nil {
+		return nil, fmt.Errorf(`relation "%s" does not exist`, cf.TableName.String())
+	}
+	insertable, ok := table.(sql.InsertableTable)
+	if !ok {
+		return nil, fmt.Errorf(`table "%s" is read-only`, cf.TableName.String())
+	}
+
+	// Open the file
+	openFile, err := os.Open(cf.File)
+	if openFile == nil || err != nil {
+		return nil, fmt.Errorf(`could not open file "%s" for reading: No such file or directory`, cf.File)
+	}
+	defer func() {
+		nErr := openFile.Close()
+		if err == nil {
+			err = nErr
+		}
+	}()
+	reader := bufio.NewReader(openFile)
+
+	// Get the row inserter and set the defers
+	rowInserter := insertable.Inserter(ctx)
+	defer func() {
+		nErr := rowInserter.Close(ctx)
+		if err == nil {
+			err = nErr
+		}
+	}()
+	rowInserter.StatementBegin(ctx)
+	defer func() {
+		if err == nil {
+			err = rowInserter.StatementComplete(ctx)
+		} else {
+			err = rowInserter.DiscardChanges(ctx, err)
+		}
+	}()
+
+	// Get the column's types, which we'll use for casting
+	sch := insertable.Schema()
+	colTypes := make([]pgtypes.DoltgresType, len(sch))
+	for i, col := range sch {
+		var ok bool
+		colTypes[i], ok = col.Type.(pgtypes.DoltgresType)
+		if !ok {
+			return nil, fmt.Errorf("COPY FROM only works for tables with all Postgres types")
+		}
+	}
+
+	// Write the data to the table
+	row := make(sql.Row, len(colTypes))
+	foundEOF := false
+	for !foundEOF {
```

Comment:
Yup, you guessed it. I debugged through this as well – `bufio.ReadLine` returns `line` and `err` when it reaches EOF and doesn't find the end of line delimiter, so you have to handle both, otherwise you'll lose data.